senders/discord/init.go

```diff
@@ -0,0 +1,98 @@
+package discord
+
+import (
+	"fmt"
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+
+	"github.com/bwmarrin/discordgo"
+	"github.com/moira-alert/moira"
+)
+
+const messenger = "discord"
+
+// Sender implements moira sender interface for discord
+type Sender struct {
+	DataBase moira.Database
+	logger   moira.Logger
+	location *time.Location
+	session  *discordgo.Session
+	frontURI string
+}
+
+// Init reads the yaml config
+func (sender *Sender) Init(senderSettings map[string]string, logger moira.Logger, location *time.Location, dateTimeFormat string) error {
+	var err error
+	token := senderSettings["token"]
+	if token == "" {
+		return fmt.Errorf("cannot read the discord token from the config")
+	}
+	sender.session, err = discordgo.New("Bot " + token)
+	if err != nil {
+		return fmt.Errorf("error creating discord session: %s", err)
+	}
+	sender.logger = logger
+	sender.frontURI = senderSettings["front_uri"]
+	sender.location = location
+
+	handleMsg := func(s *discordgo.Session, m *discordgo.MessageCreate) {
+		if err := sender.getResponse(s, m); err != nil {
+			sender.logger.Errorf("failed to handle incoming message: %s", err)
+		}
+	}
+	sender.session.AddHandler(handleMsg)
+
+	go sender.runBot()
+	return nil
+}
+
+func (sender *Sender) getResponse(s *discordgo.Session, m *discordgo.MessageCreate) error {
+
+	// Ignore all messages created by the bot itself
+	if m.Author.ID == s.State.User.ID {
+		return nil
+	}
+	// If the message is "!start" update the channel ID for the user/channel
+	if m.Content == "!start" {
+		channel, err := s.Channel(m.ChannelID)
+		if err != nil {
+			return fmt.Errorf("error while getting the channel details: %s", err)
+		}
+		switch channel.Type {
+		case discordgo.ChannelTypeDM:
+			err := sender.DataBase.SetUsernameID(messenger, "@"+m.Author.Username, channel.ID)
+			if err != nil {
+				return fmt.Errorf("error while setting the channel ID for user: %s", err)
+			}
+			msg := fmt.Sprintf("Okay, %s, your id is %s", m.Author.Username, channel.ID)
+			s.ChannelMessageSend(m.ChannelID, msg)
+		case discordgo.ChannelTypeGuildText:
+			uuid, _ := sender.DataBase.GetIDByUsername(messenger, channel.Name)
+			err := sender.DataBase.SetUsernameID(messenger, channel.Name, channel.ID)
+			if err != nil {
+				return fmt.Errorf("error while setting the channel ID for text channel: %s", err)
+			}
+			if uuid == "" {
+				msg := fmt.Sprintf("Hi, all!\nI will send alerts in this group (%s).", channel.Name)
+				s.ChannelMessageSend(m.ChannelID, msg)
+			}
+		default:
+			s.ChannelMessageSend(m.ChannelID, "Unsupported channel type")
+		}
+	}
+
+	return nil
+}
+
+func (sender *Sender) runBot() {
+	err := sender.session.Open()
```

Comment:
We must care about the simple run bot. In most Moira installations users runs many Moira notifiers (replicas), the first reason is fault tolerance (if one of your replicas stopped working you anyway has efficient service). But, for example, telegram does not allow to run many replicas of telegram bot. You must check is discord allows this? If not then just use logic like in [telegram bot](https://github.com/moira-alert/moira/blob/master/senders/telegram/init.go#L79).

Reply:
A simple way to check - just run two instances of notifier on local PC

Reply:
I tried running 2 instances of the notifier. What happens is that both instances of the bot get the events from discord and send messages to discord. Like,
![image](https://user-images.githubusercontent.com/31558262/60395122-43deb780-9b4c-11e9-9c89-305f95e4d706.png)
This is not desirable. So I think adopting the logic of the telegram bot would be the way forward.

Reply:
Yeah, I think telegam like solution is best in this situation!